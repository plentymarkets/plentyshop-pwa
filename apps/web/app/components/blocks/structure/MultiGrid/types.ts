import type { Block } from '@plentymarkets/shop-api';

export type MultiGridProps = {
  name: string;
  type: string;
  content: Block[];
  configuration: {
    columnWidths: number[];
  };
  meta: {
    uuid: string;
  };
  index?: number;
};
