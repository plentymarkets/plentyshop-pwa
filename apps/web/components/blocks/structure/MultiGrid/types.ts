import type { Block } from '@plentymarkets/shop-api';

export type MultiGridProps = {
  name: string;
  type: string;
  content: Block[];
  configuration: {
    columnWidths: number[];
  };
  meta: {
    uuid: string;
  };
  index?: number;
};

export type EmptyGridBlockProps = {
  name: string;
  type: string;
  content: [];
  meta: {
    uuid: string;
  };
};

export type AlignableBlock = Block & {
  content?: {
    imageAlignment?: string;
    alignment?: string;
    [key: string]: unknown;
  };
};

export type ColumnBlock = Block & { content?: Block[] };
