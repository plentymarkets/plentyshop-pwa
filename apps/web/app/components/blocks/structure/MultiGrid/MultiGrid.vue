<template>
  <div data-testid="multi-grid-structure" :class="getGridClasses()" :style="gridInlineStyle">
    <div
      v-for="(column, colIndex) in columns"
      :key="colIndex"
      :class="getColumnClasses(colIndex)"
      class="group/col relative overflow-hidden"
      data-testid="multi-grid-column"
    >
      <div v-for="row in column" :key="row.meta.uuid" class="group/row relative">
        <div
          v-if="showOverlay(row)"
          class="pointer-events-none absolute inset-0 opacity-0 group-hover/row:opacity-100"
          style="box-shadow: inset 0 0 0 2px #7c3aed"
        />

        <div
          v-if="showOverlay(row)"
          class="pointer-events-none absolute inset-0 z-10 opacity-0 group-hover/row:opacity-100 bg-purple-600/15"
        />

        <div
          class="absolute inset-0 z-30 flex items-center justify-center opacity-0 invisible pointer-events-none"
          :class="
            showOverlay(row)
              ? 'group-hover/row:opacity-100 group-hover/row:visible group-hover/row:pointer-events-auto'
              : ''
          "
        >
          <UiBlockActions v-if="showOverlay(row)" :block="row" :index="colIndex" :actions="getBlockActions(row)" />
        </div>

        <slot name="content" :content-block="row" />
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { ref, computed, watch, onMounted, onUnmounted, useAttrs } from 'vue';
import type { AlignableBlock, MultiGridProps } from '~/components/blocks/structure/MultiGrid/types';
import type { Block } from '@plentymarkets/shop-api';

const { itemGridHeight } = useItemGridHeight();
const { hasItemGridInColumns } = useBlockManager();
const { layout, content, configuration } = defineProps<MultiGridProps>();

const { $isPreview } = useNuxtApp();
const { isDragging } = useBlockManager();
const attrs = useAttrs() as { enableActions?: boolean; root?: boolean };
const { getSetting: getBlockSize } = useSiteSettings('blockSize');
const blockSize = computed(() => getBlockSize());

const gapClassMap: Record<string, string> = {
  None: 'gap-x-0',
  S: 'gap-y-1 md:gap-x-1 md:gap-y-0',
  M: 'gap-y-2 md:gap-x-2 md:gap-y-0',
  L: 'gap-y-3 md:gap-x-3 md:gap-y-0',
  XL: 'gap-y-5 md:gap-x-5 md:gap-y-0',
};
const gridGapClass = computed(() => gapClassMap[layout?.gap || 'M']);

const defaultMarginBottom = computed(() => {
  switch (blockSize.value) {
    case 's':
      return 30;
    case 'm':
      return 40;
    case 'l':
      return 50;
    case 'xl':
      return 60;
    default:
      return 0;
  }
});

const gridInlineStyle = computed(() => ({
  backgroundColor: layout?.backgroundColor ?? 'transparent',
  marginTop: layout?.marginTop !== undefined ? `${layout.marginTop}px` : '0px',
  marginBottom: layout?.marginBottom !== undefined ? `${layout.marginBottom}px` : `${defaultMarginBottom.value}px`,
  marginLeft: layout?.marginLeft !== undefined ? `${layout.marginLeft}px` : '40px',
  marginRight: layout?.marginRight !== undefined ? `${layout.marginRight}px` : '40px',
}));

const getGridClasses = () => {
  return gridClassFor({ mobile: 1, tablet: 12, desktop: 12 }, [gridGapClass.value ?? '', 'items-start']);
};

const getColumnClasses = (colIndex: number) => {
  const columnWidth = configuration.columnWidths[colIndex];
  return [`col-span-${columnWidth}`];
};

const enableActions = computed(() => attrs.enableActions === true);

const blockHasData = (block: Block): boolean => !!block.content && Object.keys(block.content).length > 0;

const showOverlay = computed(
  () => (block: Block) => enableActions.value && $isPreview && !isDragging.value && blockHasData(block),
);

const isAlignable = (b: Block): b is AlignableBlock =>
  typeof b.content === 'object' && b.content !== null && ('imageAlignment' in b.content || 'alignment' in b.content);

const readAlignment = (block: AlignableBlock): 'left' | 'right' | undefined => {
  const a = block.content?.imageAlignment ?? block.content?.alignment;
  return a === 'left' || a === 'right' ? a : undefined;
};

const pairWithSlots = computed<Block[]>(() => {
  const list = content.map((block) => ({ ...block }));

  const alignableIndex = list.findIndex(isAlignable);

  if (alignableIndex === -1) return list;

  const alignment = readAlignment(list[alignableIndex] as AlignableBlock);
  if (!alignment) return list;

  const selfSlot = alignment === 'right' ? 1 : 0;
  const sibling = alignableIndex === 0 ? 1 : 0;

  list[alignableIndex] = { ...list[alignableIndex], parent_slot: selfSlot } as Block;
  list[sibling] = { ...list[sibling], parent_slot: 1 - selfSlot } as Block;

  return list;
});

const columns = computed<Block[][]>(() => {
  const blocks = ref([] as Block[][]);
  pairWithSlots.value.forEach((block) => {
    if (block.parent_slot !== undefined) {
      if (!blocks.value[block.parent_slot]) {
        blocks.value[block.parent_slot] = [];
      }

      const slot = blocks.value[block.parent_slot];
      if (slot) {
        slot.push(block);
      }
    }
  });
  return blocks.value;
});

const containsItemGrid = computed(() => {
  return hasItemGridInColumns(columns.value);
});

const buttonPositionClass = ref('top-[0px]');

const baselineTop = ref(0);
const baselineScrollY = ref(0);
const currentTop = ref(0);
let ticking = false;

const MAX_DELTA = Number.POSITIVE_INFINITY;

const clamp = (n: number, min: number, max: number) => Math.min(max, Math.max(min, n));

const applyTop = () => {
  buttonPositionClass.value = `top-[${currentTop.value}px]`;
};

const handleScroll = () => {
  if (!containsItemGrid.value) return;
  if (!ticking) {
    ticking = true;
    requestAnimationFrame(() => {
      const scrollY = typeof window !== 'undefined' ? window.scrollY : 0;
      const delta = scrollY - baselineScrollY.value;
      const next = clamp(baselineTop.value + delta, baselineTop.value, baselineTop.value + MAX_DELTA);
      const rounded = Math.round(next);
      if (rounded !== currentTop.value) {
        currentTop.value = rounded;
        applyTop();
      }
      ticking = false;
    });
  }
};

const attachScroll = () => {
  baselineScrollY.value = typeof window !== 'undefined' ? window.scrollY : 0;
  currentTop.value = baselineTop.value;
  applyTop();
  if (typeof window !== 'undefined') {
    window.addEventListener('scroll', handleScroll, { passive: true });
  }
};

const detachScroll = () => {
  if (typeof window !== 'undefined') {
    window.removeEventListener('scroll', handleScroll);
  }
  currentTop.value = baselineTop.value;
  applyTop();
};

const getBlockActions = (block: Block) => {
  if (block.name === 'ItemGrid') {
    return {
      isEditable: true,
      isMovable: false,
      isDeletable: false,
      classes: ['bg-purple-400', 'hover:bg-purple-500', 'transition'],
      buttonClasses: ['border-2', 'border-purple-600', buttonPositionClass.value],
      hoverBackground: ['hover:bg-purple-500'],
    };
  }

  return {
    isEditable: true,
    isMovable: false,
    isDeletable: false,
    classes: ['bg-purple-400', 'hover:bg-purple-500', 'transition'],
    buttonClasses: ['border-2', 'border-purple-600'],
    hoverBackground: ['hover:bg-purple-500'],
  };
};
watch(
  () => itemGridHeight.value,
  (newHeight) => {
    if (containsItemGrid.value && newHeight > 0) {
      const topValue = Math.min(newHeight * 0.05, 200);
      baselineTop.value = Math.round(topValue);
      baselineScrollY.value = typeof window !== 'undefined' ? window.scrollY : 0;
      currentTop.value = baselineTop.value;
      applyTop();
    }
  },
  { immediate: true },
);

onMounted(() => {
  if (containsItemGrid.value) attachScroll();
});

onUnmounted(() => {
  detachScroll();
});

watch(
  () => containsItemGrid.value,
  (has) => {
    if (has) attachScroll();
    else detachScroll();
  },
  { immediate: true },
);
</script>
