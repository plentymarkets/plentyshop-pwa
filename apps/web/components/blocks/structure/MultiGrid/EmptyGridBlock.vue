<template>
  <div v-if="$isPreview" class="flex flex-col items-center justify-start p-10 font-editor">
    <UiButton class="mb-4 mt-20" @click.stop="addBlockToColumn()">
      <SfIconAdd class="cursor-pointer text-xl" />
    </UiButton>
    <p class="font-bold text-lg">Add block to column</p>
  </div>
</template>

<script setup lang="ts">
import { SfIconAdd } from '@storefront-ui/vue';
import type { EmptyGridBlockProps } from '~/components/blocks/structure/MultiGrid/types';

const { $isPreview } = useNuxtApp();
const props = defineProps<EmptyGridBlockProps>();
const { updateMultigridColumnUuid, visiblePlaceholder } = useBlockManager();
const { openDrawerWithView } = useSiteConfiguration();

const addBlockToColumn = () => {
  updateMultigridColumnUuid(props.meta.uuid);
  openDrawerWithView('blocksList');
  visiblePlaceholder.value = { uuid: '', position: 'top' };
};
</script>
