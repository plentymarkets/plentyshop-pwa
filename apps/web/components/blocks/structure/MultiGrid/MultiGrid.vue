<template>
  <div
    data-testid="multi-grid-structure"
    class="grid grid-cols-1 gap-4 items-center"
    :class="`lg:grid-cols-${content.length}`"
    :style="{ color: textCardBlock?.content?.text?.color || '' }"
  >
    <div v-for="(column, index) in content" :key="index" :class="`col-${configuration.columnWidths[index]}`">
      <slot name="content" :content-block="column" />
    </div>
  </div>
</template>

<script setup lang="ts">
import type { MultiGridProps } from '~/components/blocks/structure/MultiGrid/types';
import type { ImageTextProps } from '~/components/blocks/Image/types';
import type { TextCardProps } from '~/components/blocks/TextCard/types';

const { content, configuration } = defineProps<MultiGridProps>();

const imageBlock = computed(() => (content.find((block) => block.name === 'Image') || {}) as ImageTextProps);
const textCardBlock = computed(() => (content.find((block) => block.name === 'TextCard') || {}) as TextCardProps);

const swap = (arr: unknown[], from: number, to: number) => {
  arr.splice(from, 1, arr.splice(to, 1, arr[from])[0]);
};

const alignment = computed(() => imageBlock?.value?.content?.imageAlignment || 'left');

watch(
  alignment,
  () => {
    if (alignment.value === 'right' && content[0]?.name === 'Image') {
      swap(content, 0, 1);
    }
    if (alignment.value === 'left' && content[0]?.name !== 'Image') {
      swap(content, 0, 1);
    }
  },
  { immediate: true },
);
</script>
