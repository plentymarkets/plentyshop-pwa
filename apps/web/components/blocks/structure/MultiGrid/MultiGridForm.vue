<template>
  <div class="sticky top-[52px] h-[calc(100vh-52px)] overflow-y-auto">
    <div data-testid="image-text-form">
      <BlocksImageForm :uuid="imageBlock?.meta?.uuid || ''" />
      <BlocksTextCardForm :uuid="textCardBlock?.meta?.uuid || ''" />
    </div>
  </div>
</template>

<script setup lang="ts">
import type { MultiGridProps } from '~/components/blocks/structure/MultiGrid/types';
import type { TextCardProps } from '~/components/blocks/TextCard/types';
import type { ImageTextProps } from '~/components/blocks/Image/types';

const { blockUuid } = useSiteConfiguration();
const { data } = useCategoryTemplate();
const { findOrDeleteBlockByUuid } = useBlockManager();

const multiGridStructure = computed(
  () => (findOrDeleteBlockByUuid(data.value, blockUuid.value) || {}) as MultiGridProps,
);

const imageBlock = computed(
  () => (multiGridStructure.value.content.find((block) => block.name === 'Image') || {}) as ImageTextProps,
);
const textCardBlock = computed(
  () => (multiGridStructure.value.content.find((block) => block.name === 'TextCard') || {}) as TextCardProps,
);
</script>
