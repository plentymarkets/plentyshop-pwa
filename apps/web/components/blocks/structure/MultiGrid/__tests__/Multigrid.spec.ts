import { describe, it, expect } from 'vitest';
import { mount } from '@vue/test-utils';
import MultiGrid from '../MultiGrid.vue';

const multiGridBlockUuid = '22222222-2222-4222-8222-222222222222';

const mockMultiGridProps = {
  name: 'MainGrid',
  type: 'multi-grid',
  content: [
    {
      name: 'TextCard',
      type: 'text',
      content: { text: 'Test text' },
      meta: { uuid: '33333333-3333-4333-8333-333333333333' },
    },
    {
      name: 'Image',
      type: 'image',
      content: { src: '/test.jpg', alt: 'Test image' },
      meta: { uuid: '44444444-4444-4444-8444-444444444444' },
    },
  ],
  configuration: {
    columnWidths: [6, 6],
  },
  layout: {
    marginTop: 20,
    marginBottom: 20,
    marginLeft: 10,
    marginRight: 10,
    backgroundColor: '#fff',
    gap: 'M',
  },
  meta: {
    uuid: multiGridBlockUuid,
  },
};

export const mockMultiGridBlockWithContent = {
  name: 'MultiGrid',
  type: 'structure',
  meta: { uuid: '64893934-ce1b-461e-8263-da8df6753872' },
  configuration: {
    columnWidths: [6, 6],
  },
  content: [
    {
      name: 'Image',
      type: 'content',
      meta: { uuid: '45454545-4545-4455-8455-454545454545' },
      content: {
        image: {
          wideScreen: 'https://cdn02.plentymarkets.com/v5vzmmmcb10k/frontend/PWA/placeholder-image.png',
          desktop: 'https://cdn02.plentymarkets.com/v5vzmmmcb10k/frontend/PWA/placeholder-image.png',
          tablet: 'https://cdn02.plentymarkets.com/v5vzmmmcb10k/frontend/PWA/placeholder-image.png',
          mobile: 'https://cdn02.plentymarkets.com/v5vzmmmcb10k/frontend/PWA/placeholder-image.png',
          alt: 'alt text',
          imageAlignment: 'right',
          fillMode: 'fill',
          aspectRatio: '16 / 9',
        },
        text: {
          textOverlay: '',
          textOverlayColor: '#333333',
          textOverlayAlignY: 'center',
          textOverlayAlignX: 'center',
        },
        button: {
          label: '',
          link: '',
          variant: 'primary',
        },
      },
    },
    {
      name: 'TextCard',
      type: 'content',
      meta: { uuid: '56565656-5656-4565-8565-565656565656' },
      content: {
        text: {
          htmlDescription: 'Text that supports HTML formatting',
          title: 'h2 heading',
          subtitle: 'subtitle',
          textAlignment: 'left',
          color: '#000',
        },
        button: {
          label: 'Button',
          link: '/',
          variant: 'primary',
        },
        layout: {
          backgroundColor: '#ffffff',
          paddingTop: '0',
          paddingBottom: '0',
          paddingLeft: '0',
          paddingRight: '0',
          keepTransparent: 'true',
        },
      },
    },
  ],
};

describe('MultiGrid block', () => {
  it('should render the correct number of columns (2)', () => {
    const wrapper = mount(MultiGrid, {
      props: {
        name: 'MultiGrid',
        type: 'structure',
        content: [
          {
            name: 'TextCard',
            type: 'text',
            content: { text: 'Test text' },
            meta: { uuid: '33333333-3333-4333-8333-333333333333' },
          },
          {
            name: 'Image',
            type: 'image',
            content: { src: '/test.jpg', alt: 'Test image' },
            meta: { uuid: '44444444-4444-4444-8444-444444444444' },
          },
        ],
        configuration: {
          columnWidths: [6, 6],
        },
        layout: {
          marginTop: 10,
          marginBottom: 10,
          marginLeft: 5,
          marginRight: 5,
          backgroundColor: '#fff',
          gap: 'M',
        },
        meta: {
          uuid: multiGridBlockUuid,
        },
      },
    });

    const cols = wrapper.findAll('[data-testid="multi-grid-column"]');
    expect(cols.length).toBe(2);
  });

  it('should render the correct number of columns (3)', () => {
    const wrapper = mount(MultiGrid, {
      props: {
        name: 'MultiGrid',
        type: 'structure',
        content: [
          {
            name: 'TextCard',
            type: 'text',
            content: { text: 'Test text' },
            meta: { uuid: '33333333-3333-4333-8333-333333333333' },
          },
          {
            name: 'Image',
            type: 'image',
            content: { src: '/test.jpg', alt: 'Test image' },
            meta: { uuid: '44444444-4444-4444-8444-444444444444' },
          },
        ],
        configuration: {
          columnWidths: [6, 6, 6],
        },
        layout: {
          marginTop: 10,
          marginBottom: 10,
          marginLeft: 5,
          marginRight: 5,
          backgroundColor: '#fff',
          gap: 'M',
        },
        meta: {
          uuid: multiGridBlockUuid,
        },
      },
    });

    const cols = wrapper.findAll('[data-testid="multi-grid-column"]');
    expect(cols.length).toBe(2);
  });

  it('renders no columns when content is empty', () => {
    const wrapper = mount(MultiGrid, {
      props: {
        ...mockMultiGridProps,
        content: [],
      },
    });
    expect(wrapper.findAll('[data-testid="multi-grid-column"]').length).toBe(0);
  });

  it('applies the correct gap class', () => {
    const wrapper = mount(MultiGrid, {
      props: {
        ...mockMultiGridProps,
        layout: { ...mockMultiGridProps.layout, gap: 'XL' },
      },
    });
    expect(wrapper.find('[data-testid="multi-grid-structure"]').classes()).toContain('md:gap-x-5');
  });

  it('renders dynamic block types (Image, TextCard) with realistic PlentyONE content', () => {
    const wrapper = mount(MultiGrid, {
      props: mockMultiGridBlockWithContent,
    });

    const cols = wrapper.findAll('[data-testid="multi-grid-column"]');
    expect(cols.length).toBe(2);
  });

  it('applies layout styles (margin, background color) to the grid container', () => {
    const wrapper = mount(MultiGrid, {
      props: {
        ...mockMultiGridProps,
        layout: {
          marginTop: 20,
          marginBottom: 10,
          marginLeft: 5,
          marginRight: 15,
          backgroundColor: '#ABCDEF',
          gap: 'M',
        },
      },
    });

    const grid = wrapper.find('[data-testid="multi-grid-structure"]');
    const style = grid.attributes('style');
    expect(style).toContain('background-color: #ABCDEF');
    expect(style).toContain('margin: 20px 15px 10px 5px');
  });

  it('applies responsive grid classes', () => {
    const wrapper = mount(MultiGrid, {
      props: {
        ...mockMultiGridProps,
        configuration: {
          columnWidths: [4, 4, 4],
        },
      },
    });

    const grid = wrapper.find('[data-testid="multi-grid-structure"]');
    const classes = grid.classes();
    expect(classes).toContain('md:grid-cols-2');
    expect(classes).toContain('lg:grid-cols-3');
  });

  it('renders no columns when content is empty', () => {
    const wrapper = mount(MultiGrid, {
      props: {
        ...mockMultiGridProps,
        content: [],
      },
    });
    expect(wrapper.findAll('[data-testid="multi-grid-column"]').length).toBe(0);
  });
});
