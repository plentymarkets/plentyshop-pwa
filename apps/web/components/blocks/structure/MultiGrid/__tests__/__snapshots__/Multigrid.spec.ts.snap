// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MultiGrid.vue > matches snapshot 1`] = `
"<div data-testid="multi-grid-structure" class="grid gap-y-2 md:gap-x-2 md:gap-y-0 items-center grid-cols-1 md:grid-cols-2 lg:grid-cols-3" style="background-color: #fff; margin: 20px 10px;">
  <div class="group/col relative overflow-hidden">
    <!--v-if-->
    <!--v-if-->
    <div class="absolute inset-0 z-30 flex items-center justify-center opacity-0 invisible pointer-events-none">
      <!--v-if-->
    </div>
  </div>
</div>"
`;
